infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -725,7 +742,23 @@ public void append(LogAddress logAddress, LogData entry) {
             SegmentHandle fh = getSegmentHandleForAddress(logAddress);
             if (fh.getKnownAddresses().containsKey(logAddress.address) ||
                     fh.getTrimmedAddresses().contains(logAddress.address)) {
-                throw new OverwriteException();
+                if (entry.getRank()==null) {
+                    throw new OverwriteException();
+                } else {
+                    try (MultiReadWriteLock.AutoCloseableLock ignored = segmentLocks.acquireWriteLock(fh.getSegment())) {
+                        Optional<Boolean> canAppend = isAppendPermitted(logAddress, entry);
+                        if (canAppend.isPresent()) {
+                            if (canAppend.get()) {
+                                AddressMetaData addressMetaData = writeRecord(fh, logAddress.address, entry);
+                                fh.getKnownAddresses().put(logAddress.address, addressMetaData);
```

Comment:
@Maithem  you cannot, as the log address doesn't change!  But my code makes sure the trim() will still work, just we need to see how to change criteria when to be activated.

The trick is that now the code which prepares the entries to reconstruct after trimming is linked hash map by exactly this log address, not array list. So the duplicate will disappear naturally, once trim is activated.